code/grammar.txt

```diff
@@ -0,0 +1,20 @@
+
+    sentence: (verb noun conj noun prep connection | verb noun prep connection| verb noun prep noun )+
+
+
+    noun: NOUN
+    verb: VERB
+    connection: CONNECTION
+    conj: CONJUNCTION
+    prep: PREP
+
+
+    NOUN: "capacitor" | "resistor" | "battery"
+    VERB: "connect" | "add" | "remove" | "save"
+    CONNECTION: "parallel" | "series"
+    PREP: "in" | "with" | "to" | "as"
+    CONJUNCTION: "and"
+
+
+    %import common.WS
+    %ignore WS
```

Comment:
This grammar works great. I'm guessing that you probably adapted it from an existing example in the Lark documentation, which is a good idea.

However, at this stage in your project--when you're about to add more features--I suggest revising the grammar so that it uses more words associated with your domain. "Noun" and "verb" are generic language terms; you want domain-specific ones. Something like:

```
connection: "connect" components "in" CONNECTION_TYPE

// A typical grammar rule for describing a list of things. In this case, we're
// describing a list of components, separated by the word "and"
components: COMPONENT | components "and" COMPONENT

COMPONENT: "capacitor" | "resistor" | "battery"
CONNECTION_TYPE: "series" | "parallel"
```